Issue/32 (#34)
scalar-labs/scalardl-web-client-sdk: README.md

|listContracts|To list all registered contracts of the client|
|executeContract|To execute a registered contract of the client|
|validateLedger|To validate an asset of the Scalar DL network to determine if it is tampered|

> |removeCachedPrivateKey|To remove private keys store in indexedDB|

Author: The function has been renamed to `deleteIndexedDb` and moved to ClientServiceWithIndexedDb class.

};
```

## IndexedDB support
This library provides a decorator class `ClientServiceWithIndexedDb` to support storing private keys in the browsers' [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API).
We can decorate `ClientService` like

```
const clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
```

Once ClientServiceWithIndexedDb is used,
it tries to store the private key in the indexedDB or read the private key from the indexedDB depending on whether or not a private key is specified in the properties.

In the following example, tHe application tries to read ther user's private key from the indexedDB.
If the private key is not found, the application tries to get private key from an external service.

```javascript
let properties = {
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    ...
}; // Not specify 'scalar.dl.client.private_key_pem' or 'scalar.dl.client.private_key_cryptokey'

let clientService;
try {
    // try to read private key from indexedDB
    clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));

> } catch (err) {

Reviewer: We implemented two errors
- IndexedDbKeyNotFoundError
- IndexedDbOperationError
in ClientServiceWithIndexedDb

Please check it. Thank you :)

};
```

## IndexedDB support
This library supports storing private keys in the browsers' [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API).
To use the feature, please decorate `ClientService` object with `ClientServiceWithIndexedDb` as follows.

```
const clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
```

`ClientServiceWithIndexedDb` stores a private key in IndexedDB if the key is specified in client properties and reads a private key from the IndexedDB if the key is not specified in client properties.

Based on the behavior, it is recommended to use it as follows.

> If a private key is not found, `IndexedDbKeyNotFoundError` will be thrown and the application needs to get private key from an external service.

Reviewer 2: ```suggestion
If a private key is not found, `IndexedDbKeyNotFoundError` will be thrown and the application needs to get a private key from an external service.
```

```javascript
let properties = {
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    ...
}; // Not specify 'scalar.dl.client.private_key_pem' or 'scalar.dl.client.private_key_cryptokey'

let clientService;
try {
    // It tries to read a private key from IndexedDB
    clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
} catch (err) {
    if (err instanceof IndexedDbKeyNotFoundError) {
        properties['scalar.dl.client.private_key_pem'] = /* from some place */
        // This time, it stores the specified private key in IndexedDB
        clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
    } else {

>         throw err; // throw the other errors

Reviewer 2: ```suggestion
        throw err; // How to handle the error should be decided by application side
```

};
```

## Support of IndexedDB

> This library provides the support of storing private keys in the browser [indexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API).

Reviewer 3: ```suggestion
The [indexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) library provides support for storing private keys in the browser.
```

> The function can be enabled by setting `scalar.dl.client.private_key_indexeddb_enabled` to `true` in the client properties.

Reviewer 3: ```suggestion
indexedDB's functionality can be enabled by setting `scalar.dl.client.private_key_indexeddb_enabled` to `true` in the client properties.
```

> As indexedDB function is enabled, depending on whether a private key is specified in the properties,

Reviewer 3: ```suggestion
Once the indexedDB function is enabled, depending on whether or not a private key is specified in the properties,
```

> As indexedDB function is enabled, depending on whether a private key is specified in the properties,

Reviewer 2: I think we should simply explain what we should do when using IndexedDB.
I think we do the following steps in most of cases, do we ?

1. Specify `enabled: true` without `private_key_pem` specified.
2. If 1 throws error, get a private key from somewhere and specify `enabled: true` with `private_key_pem` specified with the private key.
3. If 1 doesn't throw error, go on with the key stored key in IndexedDB.

> the construction stores the private key in the indexedDB or reads the private key from the indexedDB.

Reviewer 3: ```suggestion
the construction will store the private key in indexedDB or read the private key from indexedDB.
```
In this sentence, I'm not sure what "construction" means, maybe it's not the right word?

Author: Ah, it means to execute constructor XD

> For example, with this client properties

Reviewer 3: ```suggestion
For example, with these client properties
```

```
{
    ...
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    'scalar.dl.client.private_key_pem': '-----BEGIN EC PRIVATE KEY-----\nMHc...',
    'scalar.dl.client.private_key_indexeddb_enabled': true,
    ...
}
```

> The private key will be stored in the indexedDB.

Reviewer 3: ```suggestion
The private key will be stored in indexedDB.
```

> With this client properties

Reviewer 3: ```suggestion
With these client properties
```

```
{
    ...
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    'scalar.dl.client.private_key_pem': null,
    'scalar.dl.client.private_key_indexeddb_enabled': true,
    ...
}
```

the private key will be loaded from indexedDB.
The library throws an error if the private key is not found in indexedDB.

> Notice that the private keys are stored with the index that composited by `scalar.dl.client.cert_holder_id` and `scalar.dl.client.cert_version`.

Reviewer 3: ```suggestion
Notice that the private keys are stored with an index that comprises `scalar.dl.client.cert_holder_id` and `scalar.dl.client.cert_version`.
```

> Therefore, this client properties

Reviewer 3: ```suggestion
Therefore, these client properties
```

```
{
    ...
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    'scalar.dl.client.private_key_pem': '-----BEGIN EC PRIVATE KEY-----\nMHc...',
    'scalar.dl.client.private_key_indexeddb_enabled': true,
    ...
}
```

> stores the private keys in difference index from this client properties

Reviewer 3: ```suggestion
stores the private keys in a difference index from these client properties
```

```
{
    ...
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 2,
    'scalar.dl.client.private_key_pem': '-----BEGIN EC PRIVATE KEY-----\nMHc...',
    'scalar.dl.client.private_key_indexeddb_enabled': true,
    ...
}
```

We can use `clientService.removeCachedPrivateKey()` to remove stored private keys.

> This function only removes the private key that is stored with the index composited by the the `scalar.dl.client.cert_holder_id` and `scalar.dl.client.cert_version` in the client properties.

Reviewer 3: ```suggestion
This function only removes the private keys that are stored with indexes comprising the `scalar.dl.client.cert_holder_id` and `scalar.dl.client.cert_version` in the client properties.
```
